sapi/phpdbg/phpdbg_watch.c

```diff
@@ -268,6 +305,29 @@ int phpdbg_watchpoint_segfault_handler(siginfo_t *info, void *context) {
 	return SUCCESS;
 }
 
+#ifdef HAVE_USERFAULTFD_WRITEFAULT
+void *phpdbg_watchpoint_userfaultfd_thread(void *phpdbg_globals) {
+	pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, NULL);
+	zend_phpdbg_globals *globals = (zend_phpdbg_globals *) phpdbg_globals;
+
+	struct uffd_msg fault_msg = {0};
+	while (read(globals->watch_userfaultfd, &fault_msg, sizeof(fault_msg)) == sizeof(fault_msg)) {
+    	void *page = phpdbg_get_page_boundary((char *) fault_msg.arg.pagefault.address);
+		zend_hash_index_add_empty_element(globals->watchlist_mem, (zend_ulong) page);
```

Comment:
This is problematic in a ZTS build when the table is full, and so results in allocation.